File: Interview Problems/Parking Lot/Java Code/repository/VehicleRepository.java
package repository;

import domain.Vehicle;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public class VehicleRepository {

    private final Map<UUID, Vehicle> store = new HashMap<>();

    public void save(Vehicle vehicle) {
        store.put(vehicle.getId(), vehicle);
        System.out.println("[REPO] Vehicle saved: " + vehicle.getId());
    }

    public Optional<Vehicle> findById(UUID vehicleId) {
        return Optional.ofNullable(store.get(vehicleId));
    }

    public Optional<Vehicle> findByLicensePlate(String licensePlate) {
        return store.values().stream()
                .filter(v -> v.getLicensePlate().equalsIgnoreCase(licensePlate))
                .findFirst();
    }
}

File: Interview Problems/Parking Lot/Python Code/repository/vehicle_repository.py
import uuid
from typing import Dict, Optional
from domain.vehicle import Vehicle

class VehicleRepository:
    def __init__(self):
        self._store: Dict[uuid.UUID, Vehicle] = {}

    def save(self, vehicle: Vehicle):
        self._store[vehicle.id] = vehicle
        print(f"[REPO] Vehicle saved: {vehicle.id}")

    def find_by_id(self, vehicle_id: uuid.UUID) -> Optional[Vehicle]:
        return self._store.get(vehicle_id)

    def find_by_license_plate(self, license_plate: str) -> Optional[Vehicle]:
        for v in self._store.values():
            if v.license_plate.lower() == license_plate.lower():
                return v
        return None
